### PR TITLE
DO NOT MERGE: Initial pass at JSXResourceConverter

### PR DESCRIPTION
**Pull request by ghost**
Still needs:
- [ ] more robust logging
- [x] ~~synchronize js scope access~~
- [ ] unit tests

Converts source jsx files:
image

to js files:
image

### CODE REVIEW

**Comment by ibmjon** [comment]
I'm currently converting the react POC I made over to using JSX and noticed that when there are syntax errors thrown by the transformer they're only printed in the JAGGR console (which makes sense). It would be nice to make it so that these errors get reported to the browser console some how.

**Comment by chuckdumont** [comment]
One way this could be done would be to replace the contents of the file with javascript code to log the error message in the browser when the file is loaded.

File: jaggr-core/src/main/java/com/ibm/jaggr/core/impl/resource/JsxResourceConverter.java [hunk 1]
```diff
@@ -126,7 +171,24 @@ boolean compileJsxResource(final IResource jsxRes, final File target) {
...
 			log.entering(sourceMethod, sourceMethod, new Object[]{jsxRes, target});
 		}
-		boolean result = false;	// not yet implemented
+
+		boolean result;
+		Context ctx = Context.enter();
+		try {
+			// read the contents of the jsx file and convert it
+			String jsx = IOUtils.toString(jsxRes.getInputStream());
+			NativeObject convertedJSX = (NativeObject) transformFunction.call(ctx, globalScope, jsxTransformScript, new String[]{jsx});
```
Comment (chuckdumont) [inline jaggr-core/src/main/java/com/ibm/jaggr/core/impl/resource/JsxResourceConverter.java]: Need to synchronize access to the rhino scope.  This method can be called by multiple, concurrent threads, and rhino scopes are not re-entrant.  Use synchronization to serialize access to the scope, but make sure you don't do any file I/O in the synchronized block.

File: jaggr-core/src/main/java/com/ibm/jaggr/core/impl/resource/JsxResourceConverter.java [hunk 2]
```diff
@@ -126,7 +171,24 @@ boolean compileJsxResource(final IResource jsxRes, final File target) {
...
+			// write the contents of the transformed javascript to the target file
+			String jsstring = convertedJSX.get("code").toString();  //$NON-NLS-1$
+			FileUtils.writeStringToFile(target, jsstring);
+			result = true;
+		} catch (IOException e) {
+			// TODO we should probably log something here?
```
Comment (chuckdumont) [inline jaggr-core/src/main/java/com/ibm/jaggr/core/impl/resource/JsxResourceConverter.java]: Yes.  Add logging

**Comment by chuckdumont** [comment]
The code as it currently stands will work for nocombo mode, but when using the aggregator, jsx files will not be scanned for dependencies.  This can be fixed by using the converter to convert folder resources, returning a converted resource that wraps the original folder resource and implements the walkTree() method to add .js entries for any .jsx entries encountered while walking the tree, keeping track of the synthesized entries so as to avoid calling IResourceVisitor.visitResource() more than once for the same resource if the .js file actually exists.

**Comment by ibmjon** [comment]
as part of our maven build we're going to transform all of the jsx files in to js files so that when using the aggregator it'll never hit any jsx files - do you still want us to do the dependency scanning?

**Comment by chuckdumont** [comment]
I'd still like to work in aggregator mode if you don't have all the built files, but not a high priority.

**Comment by chuckdumont** [comment]
With the current code, what happens if the maven built .js file exists in the same directory as the .jsx file and you update the jsx file.  Will the jsx file be recompiled?  I don't think so.  Should the .js file in the source location be updated in this case?  What if you don't have write access to the source directory (i.e. it's an osgi bundle)?

**Comment by cwgormanIBM** [comment]
>The code as it currently stands will work for nocombo mode, but when using the aggregator, jsx files will not be scanned for dependencies.

Why does this not happen? It seems like we are converting to .js at a very early stage (when the file is being read). Shouldn't this make the .jsx/.js conversion invisible to the rest of jaggr?

**Comment by cwgormanIBM** [comment]
> With the current code, what happens if the maven built .js file exists in the same directory as the .jsx file and you update the jsx file.

I think the plan is to not place the transpiled files in the same directory.

**Comment by chuckdumont** [comment]
It happens because the dependency scanner scans the configured paths and packages directories looking for .js files and parses them to get the dependencies that it uses to build the dependency map.  Since it only tries to open .js files that are found in a directory listing, it will never try to open a foo.js file when there is only a foo.jsx file in the directory.

**Comment by ibmjon** [comment]
>> With the current code, what happens if the maven built .js file exists in the same directory as the .jsx file and you update the jsx file.

> I think the plan is to not place the transpiled files in the same directory.

I believe the plan is to place just the transpiled files (and not their jsx counterparts) in the bundle

**Comment by cwgormanIBM** [comment]
Yeah, so in the source folder there would just be jsx files, and in the bundle there would just be js files. Right?

**Comment by ibmjon** [comment]
right.

**Comment by chuckdumont** [comment]
That won't work because when running in the development environment, the source folder and the bundle are the same thing.

**Comment by chuckdumont** [comment]
But I get your point, the jsx and transpied files will not be in the same folder.

**Comment by ibmjon** [comment]
do you have any guidance on best practices for logging?

**Comment by chuckdumont** [comment]
Use Level.WARNING to log compiler failures

**Comment by chuckdumont** [comment]
So back to the issue of jsx files not being processed by the dependency scanner; if as you say the transpiled js files won't be in the source directory and will only be in the target bundles produced by the build, then JAGGR won't be able to work unless we do something to make the transpiled js files visible to the dependency scanner.  It's no good to have this work only for nocombo.

I'll work on what we can do about this.

**Comment by chuckdumont** [comment]
@cwgormanIBM - please add the transformer files needed to run the code to this pull request.

**Comment by cwgormanIBM** [comment]
It looks like the Transformer is already included in this pull request (the first commit). 

I don't think there is a maven project for the JSX Transformer. It gets built as part of the react project: https://github.com/facebook/react

**Pull request closed without merging**